File: 96_Eular_Function.cpp
#include <iostream>
#include <cstring>

using namespace std;

long long l, r;
bool isPrime[5000001];
long long primeList[5000001];
long long phi[5000001];
long long primeCount = 0;

int main(){
    cin >> l >> r;
    memset(isPrime, true, sizeof(isPrime));
    phi[1] = 1;

    for (long long i=2; i<=r; i++){
        if (isPrime[i]){
            primeList[++primeCount] = i;
            phi[i] = i-1;
        }
        long long j=1;
        while(i*primeList[j] <= r){
            isPrime[i*primeList[j]] = false;
            if (i % primeList[j] == 0) {
                phi[i * primeList[j]] = phi[i] * primeList[j];
                break;
            }
            else
                phi[i * primeList[j]] = phi[i] * (primeList[j] - 1);
            j++;
        }
    }

    long long min = phi[l];
    long long argmin = l;
    for (long long i=l+1; i<=r; i++)
        if (phi[i] < min ){
            min = phi[i];
            argmin = i;
        }
    cout << argmin;
    return 0;
}

File: 96_Eular_Function.py
import sys

l, n = map(lambda x: int(x), raw_input().split())

isPrime = [True for i in xrange(n+1)]
primeList = [0 for i in xrange(n+1)]
phi = [1 for i in xrange(n+1)]
primeCount = 0

for i in range(2, n+1):
    if isPrime[i]:
        primeCount += 1
        primeList[primeCount] = i
        phi[i] = i - 1
    j = 1
    while i*primeList[j] <= n:
        isPrime[i * primeList[j]] = False
        if i % primeList[j] == 0:
            phi[i * primeList[j]] = phi[i] * primeList[j]
            break
        else:
            phi[i * primeList[j]] = phi[i] * (primeList[j] - 1)
        j += 1

print phi[l:].index(min(phi[l:]))+l
